apps/pkcs12.c

```diff
@@ -906,8 +906,8 @@ int print_attribs(BIO *out, const STACK_OF(X509_ATTRIBUTE) *attrlst,
             av = X509_ATTRIBUTE_get0_type(attr, 0);
             switch (av->type) {
             case V_ASN1_BMPSTRING:
-                value = OPENSSL_uni2asc(av->value.bmpstring->data,
-                                        av->value.bmpstring->length);
+                value = OPENSSL_uni2utf8(av->value.bmpstring->data,
```

Comment:
Side note: In an ideal world where everybody uses utf-8, this would be the perfect choice. Unfortunately, Windows programs often use codepage 1252 resp. iso-8859-1, where `OPENSSL_uni2asc()` produces more readable output :-/

Reply:
Conversely, in POSIX terminals that use UTF-8, `OPENSSL_uni2asc` outputs anything outside of ASCII wrong, since the terminal is expecting UTF-8. Also, on Windows, `OPENSSL_uni2asc` is still wrong because it turns codepoints outside of Latin-1 into garbage rather than escaping or replacing with some marker.

Is there a more magical function used elsewhere in the command-line tool?

Reply:
It's actually not that big problem to display UTF-8 in Windows console, `chcp 65001`. And openssl.exe can do that for you, just set OPENSSL_WIN32_UTF8 environment variable [to any value]. (Yes, it changes it back upon completion.) This suggests that if so desired, one can #ifdef _WIN32, check for availability of environment variable, and choose between 2asc and 2utf8, #endif. However, I for one would find it excessive, 2utf8 is sufficient.

Just in case for reference. Very few Unix users should have noticed the UTF8 switch in PKCS#12 in 1.1.0. On Windows situation is opposite, virtually every user would have noticed. That's where OPENSSL_WIN32_UTF8 comes into picture. Yes, it's ugly, but it's outcome of impossible choice...

Reply:
@davidben, my side note does _not_ imply that I'm opposed to your change, on the contrary. I was just expressing my frustration with using openssl in the windows command shell and raising the question, whether the improvement for linux would come at at prise for windows users. I did not know about `OPENSSL_WIN32_UTF8` (thanks for the hint, @dot-asm!). That sounds like a useful feature. I will try it next time on windows and hopefully it will reduce  my frustration :wink:.

Reply:
As for the `#ifdef _WIN32` suggestion: I would vote for following the kiss principle, i.e, don't add a hack for windows.